### PR TITLE
support optional chainId

### PR DESCRIPTION
**Pull request by hackmod**
*This is an experimental work (WIP)*

Basically this PR is based on the [net_version hack (by @turbobits)](https://github.com/ethereum/go-ethereum/pull/15002#issuecomment-394560172)

Currently, MetaMask get the chainId using `net_version` rpc call but `net_version` does not return the real `chainId`. there are no `net_chainid` nor `eth_chainid` rpc call exist. https://github.com/ethereum/go-ethereum/pull/15002

- In most case of ethereum variants, `chainId != networkId`.
- This is a **quick hack** to support optional chainid in the custom Rpc.
- experimental **ETC** support added

## Done
- [x] fixed custom Rpc form to support optional chainid.
  - [x] support both old-ui and new-ui
- [x] experimental ETC support added
  - basically this is a predefined `rpcUrl` + `chainId`
- [x] ticker (6/21)
- [x] block explorer urls (6/21)
- [x] support shapeshift exchange for deposit(6/22)
- [x] support testnet faucet (6/23)
  - ~~[x] MORDEN testnet added~~
- [x] more configurable settings like as "blockExplorerUrl"
  - rpcUrl, chainId, blockExplorerUrl, ticker symbol
- [x] increased customRpcList size from 2 to 3

## ToDo
- [ ] QA test / fix possible bugs
- [ ] sanity check rpcUrl
- ~~[x] FIXME morden api server url~~

## References
- https://github.com/MetaMask/metamask-extension/issues/1820
- https://github.com/ethereum/go-ethereum/pull/15002#issuecomment-394560172

## Screenshot

image

image

image

image

image

image

### CODE REVIEW

**Comment by shd101wyy** [comment]
This pull request is awesome. Thanks a lot. I also have the demand to set the chainId.
Hope the maintainer can review your pull request soon. Good job.

File: ui/lib/account-link.js [hunk 1]
```diff
@@ -18,6 +18,7 @@ module.exports = function (address, network) {
       link = `https://kovan.etherscan.io/address/${address}`
       break
     default:
+      // FIXME
```
Comment (wjmelements) [inline ui/lib/account-link.js]: Does [GasTracker.io](https://gastracker.io/) work for this purpose?

**Comment by wjmelements** [comment]
Infura doesn't support ETC, so you would need to provide a default RPC provider or enforce a custom RPC URL for non-Infura networks.

File: ui/lib/account-link.js (hunk 1 above)
Comment (splix) [reply ui/lib/account-link.js]: yes, would be `https://gastracker.io/addr/${address}`

**Comment by splix** [comment]
Does Infura provides something specific for metamask, or standard web3 JSON RPC would work? If later works, there is one at https://web3.gastracker.io

File: old-ui/app/config.js [hunk 2]
```diff
@@ -249,6 +269,11 @@ function currentProviderDisplay (metamaskState) {
       value = 'Rinkeby Test Network'
       break
 
+    case 'classic':
+      title = 'Current Network'
+      value = 'Ethereum Classic Network'
+      break
```
Comment (wjmelements) [inline old-ui/app/config.js]: Similarly, you could add the ETC Testnet, [Morden](https://github.com/ethereumclassic/classic-guide/blob/master/source/network/test-networks.rst).

**Comment by wjmelements** [comment]
I'm surprised information about a network is not already centralized in an object. That might be a good refactor, though perhaps outside the scope of this pull request.

**Comment by kumavis** [comment]
Another issue here is showing conversion rate
we're still working on breaking out networks for first class treatment which will enable non-evm blockchains as well, but longer time line

**Comment by hackmod** [comment]
screenshot updated
- support block explorer
- current price using https://min-api.cryptocompare.com/ (mist method)
- etc logo added for beta UI

**Comment by pyskell** [comment]
BUG: Doesn't support sending transaction data.

Tried sending a transaction with: `0x657463206f6e206d6574616d61736b20697320776f726b696e6721`
image

Result:
image

**Comment by hackmod** [comment]
This is a new UI testing result.

https://gastracker.io/tx/0x44fce132106162045005ee5815bffc0b95a5a8b6845273f4fbda2e5f5d97c7cb

image

File: app/scripts/controllers/network/networks.js [hunk 3]
```diff
@@ -18,6 +20,18 @@ networks.networkList = {
...
     'buyUrl': '',
   },
+  [MODERN]: {
+    'chainId': MODERN_CODE,
+    'ticker': 'ETC',
+    'blockExplorerTx': 'https://gastracker.io/tx/[[txHash]]',
+    'blockExplorerAddr': 'https://gastracker.io/addr/[[address]]',
+    'blockExplorerToken': 'https://gastracker.io/token/[[tokenAddress]]/[[address]]',
+    'service': 'Ethereum Commonwealth',
+    'rpcUrl': 'https://etc-geth.0xinfra.com',
```
Comment (hackmod) [inline app/scripts/controllers/network/networks.js]: where can I find api server for the modern testnet?

File: app/scripts/controllers/network/networks.js [hunk 4]
```diff
@@ -18,6 +20,18 @@ networks.networkList = {
...
     'buyUrl': '',
   },
+  [MORDEN]: {
+    'chainId': MORDEN_CODE,
+    'ticker': 'ETC',
+    'blockExplorerTx': 'https://gastracker.io/tx/[[txHash]]',
+    'blockExplorerAddr': 'https://gastracker.io/addr/[[address]]',
+    'blockExplorerToken': 'https://gastracker.io/token/[[tokenAddress]]/[[address]]',
+    'service': 'Gastracker.io',
+    'rpcUrl': 'https://web3.gastracker.io/morden',
```
Comment (hackmod) [inline app/scripts/controllers/network/networks.js]: https://web3.gastracker.io/morden is not work correctly with MetaMask.
@pyskell // is there any official api-server/block explorer exist for the Morden testnet ?

File: app/scripts/controllers/network/networks.js [hunk 5]
```diff
@@ -18,6 +20,18 @@ networks.networkList = {
...
     'buyUrl': '',
   },
+  [MORDEN]: {
+    'chainId': MORDEN_CODE,
+    'ticker': 'ETC',
+    'blockExplorerTx': 'http://mordenexplorer.ethertrack.io/tx/[[txHash]]',
+    'blockExplorerAddr': 'http://mordenexplorer.ethertrack.io/addr/[[address]]',
+    'blockExplorerToken': 'http://mordenexplorer.ethertrack.io/token/[[tokenAddress]]/[[address]]',
+    'service': 'Gastracker.io',
+    'rpcUrl': 'https://etc-geth.0xinfra.com',
```
Comment (hackmod) [inline app/scripts/controllers/network/networks.js]: `FIXME`  is there any official api-server/block explorer exist for the Morden testnet ?

Comment (wjmelements) [reply app/scripts/controllers/network/networks.js]: Maybe contact the ETC rpc providers about adding support for the testnet. If there isn't a reliable rpc provider for the ETC testnet, don't list it among the supported networks.

Comment (akx20000a) [reply app/scripts/controllers/network/networks.js]: @hackmod Remove modern testnet, no one is using ethereum classic testnet for testing purpose. Why do we need the option??

Comment (hackmod) [reply app/scripts/controllers/network/networks.js]: OK. as you already know, this is a quick hack and I'm not much familiar with a react program.
we need more testing to find possible bugs. that's the reason the morden testnet added.

**Comment by hackmod** [comment]
rebased with small conflict fix

File: app/scripts/controllers/network/networks.js (hunk 5 above)
Comment (hackmod) [reply app/scripts/controllers/network/networks.js]: temporary added morden api server.
but the api server provided by other coin project  seems improper to me.

Comment (eosclassicteam) [reply app/scripts/controllers/network/networks.js]: @hackmod Then remove useless morden testnet menu, we've added the endpoint only for this PR merge if you don't need it we will close then

**Comment by hackmod** [comment]
**current screenshot**

image

image

image

image

File: app/scripts/controllers/network/networks.js (hunk 5 above)
Comment (hackmod) [reply app/scripts/controllers/network/networks.js]: removed and rebased

**Comment by hackmod** [comment]
more configurable option for other ethereum networks or private network

image

**Comment by hackmod** [comment]
- [x] increased customRpcList size from 2 to 3
image

**Comment by akx20000a** [comment]
야 제정신이냐

**Comment by wjmelements** [comment]
아니

**Comment by akx20000a** [comment]
@wjmelements @hackmod 아니긴 뭐가 아니야 씨발년아 이 틀딱 관종새끼가 업데이트 안하고 개 밍기적 대는거 내가 빨리 고쳐서 쓸려고 하니까 존나 딴지걸고 지랄하는데 정작 지 새끼는 딴 애들이 공개한거 쓰는데 아무 거리낌이 없으면서 내가 쓰고 기여하고 공개하려고 하면 존나 지랄하네 너 씨발 ESN 레파지토리에다가 multi-geth 랑 내꺼 사용하는지 유심히 보고 있다 씨발년아 또 내 PR 와서 지랄 똥 싸면 대림동 니 집 앞에 똥칠한다 씹새끼야

**Comment by akx20000a** [comment]
tlqkf

**Comment by PeterTheOne** [comment]
I've tried it with a custom RPC and ticker symbol, seems to work fine, didn't encounter any bugs. keep up the good work!

**Comment by danfinlay** [comment]
This is nice work, but we are not ready to add a bunch of new networks and providers to our menu. If all this PR did was add an optional chainId parameter, we would merge it much more easily, but it instead adds many other things as well, which we are simply not ready to do at this time.

You can read more about our roadmap and vision for supporting multiple networks here:
https://medium.com/metamask/metamasks-vision-for-multiple-network-support-4ffbee9ec64d

**Comment by hackmod** [comment]
>  bunch of new networks and providers to our menu

@danfinlay
I guess you miss understood this PR

the following screen shot is mere *an example*. all question mark prefixed items are configured by user by means of settings menu.
image

image

and *the only newly added menu item is ETC*.

optional chain_id feature and pre-configured menu(ETC) shares their code, that's why this PR named as "optional chain Id"

anyway, I will separate this issue into several PRs, for example,
- optional chain id
- more configurable settings for custom Rpc
- ETC support

**Comment by irthos** [comment]
Hey guys and gals. Everything looks good. Anything we still need to help out with to get it shipped with ETC cupport?

**Comment by bdresser** [comment]
@irthos we've moved to https://github.com/MetaMask/metamask-extension/pull/5134 which we'll review this week once we get https://github.com/MetaMask/metamask-extension/issues/4510 merged!

**Comment by irthos** [comment]
Thank u, big ups!
